tests/test_upload.py

```diff
@@ -694,6 +694,300 @@ def test_init_object_write_grpc_final_message_empty_data(self):
+                resource={"name": "object", "bucket": "projects/_/buckets/bucket-name"},
+            )
+            request = storage_pb2.BidiWriteObjectRequest(
+                write_object_spec=spec,
+                write_offset=0,
+                object_checksums=test["checksums"],
+                finish_write=True,
+            )
+
+            context = self.mock_context()
+            db = unittest.mock.Mock()
+            db.get_bucket = unittest.mock.MagicMock(return_value=bucket)
+            responses = list(
+                gcs.upload.Upload.process_bidi_write_object_grpc(db, [request], context)
+            )
+            # Verify the annotations inserted by the testbench.
+            annotations = responses[0].resource.metadata
+            expected = test["expected"]
+            self.maxDiff = None
+            self.assertEqual(
+                annotations,
+                {**annotations, **expected},
+                msg="Testing with %s checksums" % name,
+            )
+
+    def test_process_bidi_write_grpc_empty(self):
+        db = unittest.mock.Mock()
+        context = self.mock_context()
+        list(gcs.upload.Upload.process_bidi_write_object_grpc(db, [], context))
+        context.abort.assert_called_once_with(
+            grpc.StatusCode.INVALID_ARGUMENT, unittest.mock.ANY
+        )
+
+    def test_process_bidi_write_grpc_final_message_empty_data(self):
+        request = testbench.common.FakeRequest(
+            args={}, data=json.dumps({"name": "bucket-name"})
+        )
+        bucket, _ = gcs.bucket.Bucket.init(request, None)
+        request = storage_pb2.StartResumableWriteRequest(
+            write_object_spec=storage_pb2.WriteObjectSpec(
+                resource={"name": "object", "bucket": "projects/_/buckets/bucket-name"}
+            )
+        )
+        context = self.mock_context()
+        upload = gcs.upload.Upload.init_resumable_grpc(
+            request, bucket.metadata, context
+        )
+
+        line = b"The quick brown fox jumps over the lazy dog"
+        r1 = storage_pb2.BidiWriteObjectRequest(
+            upload_id=upload.upload_id,
+            write_offset=0,
+            checksummed_data=storage_pb2.ChecksummedData(
+                content=line, crc32c=crc32c.crc32c(line)
+            ),
+            finish_write=False,
+        )
+        r2 = storage_pb2.BidiWriteObjectRequest(
+            write_offset=len(line),
+            finish_write=True,
+        )
+        context = self.mock_context()
+        db = unittest.mock.Mock()
+        db.get_bucket = unittest.mock.MagicMock(return_value=bucket)
+        db.get_upload = unittest.mock.MagicMock(return_value=upload)
+        responses = list(
+            gcs.upload.Upload.process_bidi_write_object_grpc(db, [r1, r2], context)
+        )
+        blob = responses[0].resource
+        self.assertEqual(blob.name, "object")
+        self.assertEqual(blob.bucket, "projects/_/buckets/bucket-name")
+
```

Comment:
Consider a test for "invalid data field":

https://app.codecov.io/gh/googleapis/storage-testbench/pull/586?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=googleapis#83ee5f7e84364b0abad9908f111e7409-R335

Reply:
On second thought, we probably don't need this invalid field check after all. I think in most scenarios we can rely on protobuf message checking - for a oneof group, checks if any field inside is set and raises ValueError for anything not defined in the message descriptor.